### PR TITLE
refactor: Use DiffResult all the way

### PR DESCRIPTION
**Pull request by arv**
Instead of having a `string[]` with the changed keys we now use
`DiffResult<ReadonlyJSONValue>[]` all the way.

TODO: This does not rename things yet.

Towards #839

### CODE REVIEW

File: src/sync/pull.ts [hunk 1]
```diff
@@ -245,10 +246,12 @@ export type ReplayMutation = {
 // "" is used for the primary index.
 export type ChangedKeysMap = Map<string, string[]>;
 
+export type ChangedDiffsMap = Map<string, DiffResult<ReadonlyJSONValue>[]>;
```
Comment (grgbkr) [inline src/sync/pull.ts:249]: ChangedDiffs seems redundant.  Perhaps DiffsMap, or DiffResultsMap.

File: src/db/write.ts [hunk 2]
```diff
@@ -264,30 +265,28 @@ export class Write extends Read {
...
     const valueHash = await this.map.flush();
-    let valueChangedKeys: string[] = [];
-    if (generateChangedKeys && this._basis) {
+    let valueDiff: DiffResult<ReadonlyJSONValue>[] = [];
+    if (generateDiffs && this._basis) {
       const basisMap = new BTreeRead(this._dagWrite, this._basis.valueHash);
-      valueChangedKeys = await asyncIterableToArray(
-        this.map.diffKeys(basisMap),
-      );
+      valueDiff = await asyncIterableToArray(this.map.diff(basisMap));
```
Comment (grgbkr) [inline src/db/write.ts:280]: should this import diff from diff.ts and use that?

File: src/db/write.ts [hunk 3]
```diff
@@ -296,20 +295,24 @@ export class Write extends Read {
...
-              asyncIterableToArray(map.diffKeys(basisMap)),
-            );
-          }
-          return asyncIterableToArray(map.keys());
-        },
-      );
+      const indexDiffResult = await index.withMap(this._dagWrite, async map => {
+        if (basisIndex) {
+          return basisIndex.withMap(this._dagWrite, basisMap =>
+            asyncIterableToArray(map.diff(basisMap)),
```
Comment (grgbkr) [inline src/db/write.ts:301]: should this use diff from diff.ts?

File: src/sync/pull.ts [hunk 4]
```diff
@@ -405,18 +408,26 @@ function assertResult(v: any): asserts v is Result {
 
   assertHTTPRequestInfo(v.httpRequestInfo);
 }
+
 async function addChangedKeysForIndexes(
```
Comment (grgbkr) [inline src/sync/pull.ts:412]: addChangedKeysForIndexes  => addDiffsForIndexes

File: src/sync/pull.ts [hunk 5]
```diff
@@ -427,18 +438,18 @@ async function addChangedKeysForIndexes(
       const newIndex = newIndexes.get(oldIndexName);
       if (newIndex !== undefined) {
         const changedKeys = await newIndex.withMap(read, async newMap => {
```
Comment (grgbkr) [inline src/sync/pull.ts:440]: changedKeys => diffs

File: src/sync/pull.ts [hunk 6]
```diff
@@ -427,18 +438,18 @@ async function addChangedKeysForIndexes(
...
         newIndexes.delete(oldIndexName);
         if (changedKeys.length > 0) {
-          changedKeysMap.set(oldIndexName, changedKeys);
+          diffsMap.set(oldIndexName, changedKeys);
         }
       } else {
-        // old index name is not in the new indexes. All keys changed!
-        const changedKeys = await allKeys(oldMap);
+        // old index name is not in the new indexes. All entries changed!
+        const changedKeys = await diffForAll(oldMap);
```
Comment (grgbkr) [inline src/sync/pull.ts:450]: changedKeys => diffs

File: src/sync/pull.ts [hunk 7]
```diff
@@ -447,9 +458,9 @@ async function addChangedKeysForIndexes(
   for (const [newIndexName, newIndex] of newIndexes) {
     // new index name is not in the old indexes. All keys changed!
     await newIndex.withMap(read, async newMap => {
-      const changedKeys = await allKeys(newMap);
+      const changedKeys = await diffForAll(newMap);
```
Comment (grgbkr) [inline src/sync/pull.ts:461]: changedKeys => diffs

File: src/sync/pull.ts [hunk 8]
```diff
@@ -447,9 +458,9 @@ async function addChangedKeysForIndexes(
...
       if (changedKeys.length > 0) {
-        changedKeysMap.set(newIndexName, await allKeys(newMap));
+        diffsMap.set(newIndexName, await diffForAll(newMap));
```
Comment (grgbkr) [inline src/sync/pull.ts:463]: We don't need to compute diffForAll again, we can just use the value computed above (currently named changedKeys, but should be named diffs).

File: src/sync/pull.ts (hunk 6 above)
Comment (grgbkr) [inline src/sync/pull.ts:450]: This needs to be all deleted, not all added, correct?  I think diffForAll need to take an op.

Comment (grgbkr) [reply src/sync/pull.ts:450]: Seems we probably need a test for this.

**Review by grgbkr — APPROVED** [review]

File: src/db/write.ts [hunk 9]
```diff
@@ -296,20 +295,24 @@ export class Write extends Read {
     for (const [name, index] of this.indexes) {
```
Comment (grgbkr) [inline src/db/write.ts:295]: I think this is missing handling for indexes that are in basisIndexes but missing from this.indexes.

File: src/sync/pull.ts (hunk 6 above)
Comment (arv) [reply src/sync/pull.ts:450]: I see what you are saying now... Will look

File: src/db/write.ts (hunk 9 above)
Comment (arv) [reply src/db/write.ts:295]: Not sure. Filing an issue to not forget this

Comment (arv) [reply src/db/write.ts:295]: You are correct. This is not dealing with that case. Good catch.

**Pull request merged**